tomlkit/parser.py

```diff
@@ -474,6 +474,9 @@ def _parse_value(self):  # type: () -> Item
         with self._state:
             return self._parse_literal_string()
 
+        with self._state:
```

Comment:
Initializing a state before even knowing if we have a potential array will impact performance.

I think we can still have the array parsing in its own method but the character check should still be in `_parse_value()`.

Reply:
I agree I've seen a slight slowdown. My ultimate desire though was to get `_parse_value` to look like this:

```python
def _parse_value(self):
    for parser in self._value_parsers:
        with self.state:
            return parser()
```

where `self._value_parsers` is the various parsing methods. Maybe it should look something more like this then:

```python
def _parse_value(self):
    for parser in self._value_parsers:
        try:
            return parser()
        except Restore:
            pass
```

where a `self._value_parsers` looks like this:

```python
def _parse_array(self):
    if self._current != "[":
        raise Restore

    with self.state:
        ...
```

That way we are still moving towards modularity but also have the ability to exit a scope before a state is made?